test/integration/theme-schema.test.js

```diff
@@ -45,4 +49,13 @@ describe( 'theme.json schema', () => {
 
 		expect( result ).toBe( true );
 	} );
+
+	test.each( invalidFiles )( 'invalidates schema for `%s`', ( filepath ) => {
```

Comment:
I would appreciate it if you could tell me why it is necessary to validate invalid `theme.json` files.

Isn't it enough to simply check whether `theme.json` files in the Gutenberg project are valid via the `validates schema for %s` test? 🤔

Reply:
> I would appreciate it if you could tell me why it is necessary to validate invalid `theme.json` files.
> 
> Isn't it enough to simply check whether `theme.json` files in the Gutenberg project are valid via the `validates schema for %s` test? 🤔

Good question! Testing should have a good purpose, and I don't think it always does.

In software development this kind of testing is called [_negative testing_](https://en.wikipedia.org/wiki/Negative_testing). It's covering failure paths to make sure that they still fail in a way that you would expect.

The theme.json schema is used by many developers, not just core. We want to ensure that invalid properties are flagged for those theme developers as they are developing their themes.

![Invalid properties in theme.json are underlined](https://developer.wordpress.org/news/files/2024/07/schema-validation-vscode-no-shadow.png)

In my personal philosophy of testing, I see tests as a communication tool much like short self-contained correct "compileable" examples ([SSCCEs](http://www.sscce.org/)). It can help people who are less familiar with a change understand specifically what has been changed.

The theme.json schema is a complicated schema with a lot of `anyOf`/`allOf` logic and `$ref`s. In this update, I manually changed the way that we invalidate some of those properties by using `propertyNames` instead of `additionalProperties`, and it isn't immediately obvious where each of those changes affects validation in theme.json. It actually took me a while to trace through all the references in the schema to generate those test files.

As a bonus, tests like this can reduce the burden of manual testing by reviewers when there are a lot of cases to test. Instead of going through each case, if the test can be written simply enough, it's often easier to check the validity of a single test than many manual cases.

I've written the test cases so that they should pass for the old schemas and the new schemas. If one of the tests passed with the old schemas and failed with the new schemas, we would know that I messed something up in my refactor since these properties should be invalid in both cases.